[Experimental] Product Filters: add layout and block spacing support

dinhtungdu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR adds the Block Spacing setting to the Product Filters block by adding `blockGap` support. `layout` support is required for block spacing to work, so it was added in this PR too. The layout setting is based on the following design:

![image](https://github.com/woocommerce/woocommerce/assets/5423135/d092b8b0-8e20-4425-801a-e5cc1dca2786)


<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #48022.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR should be tested by developers only.

0. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester Plugin, you can enable the experimental blocks in Settings > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.
2. Edit Block Spacing in the Editor:
    - Go to Site Editor > Templates > Product Catalog, add the "Product Filters (Experimental)" wrapper block to the template.
    - See Layout settings and Block Spacing settings in the sidebar.
    - Adjust the block spacing setting.
    - Verify that the spacing changes are immediately visible in the editor.
1. Verify Spacing on Frontend:
    - Save the changes and view the page or post on the frontend.
    - Confirm that the spacing adjustments made in the editor are accurately reflected on the frontend.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

dinhtungdu committed:
chore: changelog

nefeline reviewed (commented):
Thanks for working on this, Tung!

A few recommendations and questions:

- Could you please update the test instructions to ensure that whoever is testing this PR, has the "experimental blocks" feature enabled in the WooCommerce Beta Tester plugin? If they don't, the **Product Filters (Experimental)** block won't be available for insertion in the Product Catalog.

Here are a few screenshots of the results I got while testing this PR, both on the editor and on the frontend:

<img width="1229" alt="Screenshot 2024-06-13 at 14 22 32" src="https://github.com/woocommerce/woocommerce/assets/15730971/ec39b90b-424f-42b4-be0a-f9031175e1f6">

<img width="1229" alt="Screenshot 2024-06-13 at 14 22 22" src="https://github.com/woocommerce/woocommerce/assets/15730971/631879d7-eb2c-46b5-aa31-b4f1b1296db6">


The Block Spacing setting works as expected, but to confirm: is this the expected outcome in terms of design?

dinhtungdu committed:
add: vertical aligment controls

dinhtungdu commented:
@nefeline thank you for your review, I updated the testing instructions mentioning experimental blocks. About the design, I changed the default vertical align to `top`, so the horizontal layout should look better now.

<img width="1210" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5423135/9083f1fa-dd0c-4b10-be40-43e23ed0f0a1">

I had a conversation with @jarekmorawski on wrapping, he said that we want wrapping to support smaller screens. It also looks better with top alignment, but I don't like the long attribute list. What do you think Jarek? Happy to follow any direction here to make horizontal layout look good by default.

<img width="1138" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5423135/18e31d4b-1a2a-4758-83f9-6816600b74d2">

jarekmorawski commented:
> About the design, I changed the default vertical align to top, so the horizontal layout should look better now.

Thanks, @dinhtungdu. I confirm this should be the default setting. 🙏 

> It also looks better with top alignment, but I don't like the long attribute list. What do you think Jarek? Happy to follow any direction here to make horizontal layout look good by default.

Honestly, there's no way to make it look good without introducing dropdowns or accordions (currently WIP). We can, however, offer truncation if the list of attributes grows too long. I planned to make it a future improvement to limit the scope for the MVP, but feel free to take this on if you have space.

> If the attribute has more than 15 terms, we add a “Show more” button. It has the same styling as the “Clear” button. Users can’t click it in the canvas, but it will allow shoppers to expand the list of available terms on the frontend.

<img width="553" alt="image" src="https://github.com/woocommerce/woocommerce/assets/79307566/be0c6ad7-6210-43aa-848d-b36436284d31">

dinhtungdu committed:
add: layout and block spacing support

thealexandrelara reviewed (commented):
Thank you for working on this @dinhtungdu. I just tested and here are the results:

✅ Merchants can justify items to the left, right, center or stretch items.
✅ Merchants can change the inner blocks orientation from vertical to horizontal.
⚠️ When the orientation is set to horizontal, the "Active" filters heading is not aligned to the top like the others:
![CleanShot 2024-06-18 at 14 53 50@2x](https://github.com/woocommerce/woocommerce/assets/20469356/5135e61e-2ba3-46b2-9248-d5e73e8471bb)
⚠️ Since you were unable to join our last team meetup, I believe you're not aware of one of the things we discussed about always including E2E tests in our PRs. Would you mind including some E2E tests? For example, one test suggestion would be to use [CSS to match elements based on layout](https://playwright.dev/docs/other-locators#css-matching-elements-based-on-layout) to detect if the blocks appear horizontally when the orientation is set to horizontal.

dinhtungdu commented:
> When the orientation is set to horizontal, the "Active" filters heading is not aligned to the top like the others:

As I mentioned [above](https://github.com/woocommerce/woocommerce/pull/48429#issuecomment-2167435040), in the new design, the Active Filters shouldn't have a title by default. I didn't touch that block in this PR because it's out of scope.

> always including E2E tests in our PRs

Love that. I will add some for this PR.

dinhtungdu committed:
test: ensure the block spacing setting is visible

dinhtungdu committed:
test: ensure the layout settings is rendered

dinhtungdu committed:
test: default layout settings

dinhtungdu committed:
test: justification setting

dinhtungdu committed:
test: orientation

dinhtungdu committed:
test: block spacing

dinhtungdu committed:
fix: update test to check for inspector styles controls

dinhtungdu committed:
test: remove unnecessary util

dinhtungdu committed:
test: store selector in blockData

dinhtungdu commented:
@thealexandrelara I add some tests to verify new settings. This PR is ready for another review, can you please give it another look?

dinhtungdu commented on plugins/woocommerce-blocks/tests/e2e/tests/product-filters/product-filters.block_theme.spec.ts:
```diff
@@ -158,5 +162,148 @@ test.describe( `${ blockData.name }`, () => {
+		await expect(
+			layoutSettings.getByLabel( 'Space between items' )
+		).toBeVisible();
+		await expect(
+			block.locator( blockData.selectors.editor.layoutWrapper )
+		).toHaveCSS( 'flex-direction', 'row' );
```
I'm not happy with CSS locator and assertion, but it seems to be the only way to test these settings AFAIK. I'm open to any suggestions.

thealexandrelara replied:
One option would be to use css locators based on layout position, like in these [examples](https://playwright.dev/docs/other-locators#css-matching-elements-based-on-layout):

```
// Fill an input to the right of "Username".
await page.locator('input:right-of(:text("Username"))').fill('value');

// Click a button near the promo card.
await page.locator('button:near(.promo-card)').click();

// Click the radio input in the list closest to the "Label 3".
await page.locator('[type=radio]:left-of(:text("Label 3"))').first().click();
```

dinhtungdu replied:
Ah, I checked your [suggestion](https://github.com/woocommerce/woocommerce/pull/48429#pullrequestreview-2126170020) but I haven't gotten it until now. We can check if the text Status which is right of Price is visible or not to verify the layout direction. Thank you, I will try it in this PR.

About the block spacing tests, do you have any alternative?

thealexandrelara reviewed (approved):
Great work! I left a comment regarding the CSS locators in Playwright, but it is not a blocker for this PR, so I'm pre-approving it!

dinhtungdu committed:
test: use layout matching
